File: Algorithms/Tree and Graph/DFS.swift
//
//  DFS.swift
//  Algorithms
//
//  Created by Rahul Ranjan on 3/29/17.
//  Copyright © 2017 Rahul Ranjan. All rights reserved.
//

import Foundation

typealias VisitedDictionary = [Vertex<String>: Bool]

func depthFirstSearch(graph: AdjancencyListGraph<String>, source: Vertex<String>) -> [Vertex<String>] {
    
  var stack = Stack<Vertex<String>>()

  var visited: VisitedDictionary = {
    var v = VisitedDictionary()
    graph.adjacencyDict.keys.forEach {
      v[$0] = false
    }
    return v
  }()

  var path: [Vertex<String>] = []

  // Add the first node and mark it visited
  stack.push(source)
  visited[source] = true
    
  while !stack.isEmpty {
    let topElement = stack.pop()!
    path.append(topElement)
    visited[topElement] = true
    
    // get the neighbours
    if let neigbhours = graph.neighbours(for: topElement) {
      for neigbhour in neigbhours {
        if visited[neigbhour] == false {
          stack.push(neigbhour)
        }
      }
    }
  }
  return path
}

func depthFirstSearchTest() {
  let (graph, source) = initialiseAdjacencyListGraph()
  let path = depthFirstSearch(graph: graph, source: source)

  print(path)
}

File: Algorithms/Python/Graph/DFS.py
#!/usr/bin/python3


def DFSUtil(s, g, v, p):
	if v[s]: return
	
	v[s] = 1
	p.append(s)
	
	for n in g[s]:
		if v[n] == 0:
			DFSUtil(n, g, v, p)
	
def DFS(source, graph):
	v = { k:0 for k in graph.keys() }
	path = list()
	DFSUtil(source, graph, v, path)
	return path
	
def testDFS():
	graph = {
		1: [2, 4],
		2: [3, 1, 5],
		3: [2, 5],
		5: [2, 3],
		4: [1]
	}
	
	path = DFS(1, graph)
	print(path)

if __name__ == "__main__":
	testDFS()